File: 08. Trees/01. Preorder Traversal/main.py
class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

# Preorder Traversal (Recursive)
def preorder_traversal_recursive(root, result):
    if not root:
        return
    result.append(root.val)
    preorder_traversal_recursive(root.left, result)
    preorder_traversal_recursive(root.right, result)

# Preorder Traversal (Iterative)
def preorder_traversal_iterative(root):
    if not root:
        return []
    
    stack, result = [root], []

    while stack:
        node = stack.pop()
        result.append(node.val)

        # Push right first so that left is processed first
        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)

    return result

# Main function
def main():
    # Sample tree:   1
    #              /   \
    #             2     3
    #            / \   / \
    #           4   5 6   7

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(7)

    recursive_result = []
    preorder_traversal_recursive(root, recursive_result)
    print("Recursive Preorder Traversal:", recursive_result)

    iterative_result = preorder_traversal_iterative(root)
    print("Iterative Preorder Traversal:", iterative_result)

# Run the main function
main()

File: 08. Trees/01. Preorder Traversal/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition of TreeNode
class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;
    
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// Preorder Traversal (Recursive)
void preorderTraversalRecursive(TreeNode* root, vector<int>& result) {
    if (!root) return;
    result.push_back(root->val);
    preorderTraversalRecursive(root->left, result);
    preorderTraversalRecursive(root->right, result);
}

// Preorder Traversal (Iterative)
vector<int> preorderTraversalIterative(TreeNode* root) {
    vector<int> result;
    if (!root) return result;

    stack<TreeNode*> st;
    st.push(root);

    while (!st.empty()) {
        TreeNode* node = st.top();
        st.pop();
        result.push_back(node->val);

        // Push right first so that left is processed first
        if (node->right) st.push(node->right);
        if (node->left) st.push(node->left);
    }
    return result;
}

int main() {
    // Sample tree:   1
    //              /   \
    //             2     3
    //            / \   / \
    //           4   5 6   7

    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(7);

    // Recursive Preorder Traversal
    vector<int> recursiveResult;
    preorderTraversalRecursive(root, recursiveResult);
    cout << "Recursive Preorder Traversal: ";
    for (int val : recursiveResult) cout << val << " ";
    cout << endl;

    // Iterative Preorder Traversal
    vector<int> iterativeResult = preorderTraversalIterative(root);
    cout << "Iterative Preorder Traversal: ";
    for (int val : iterativeResult) cout << val << " ";
    cout << endl;

    return 0;
}
